overlays: fix osk placeholder and add password mode

Megamouse opened the pull request:
fixes #10100

- Adds a value member to the overlay text edit in order to seperate the visible text from the actual text value
- Uses the new value member to implement a proper placeholder logic (when the text edit is empty)
- Uses the new value to implement the password mode (visible text is displayed as a string of ＊)
- Adds the password echo mode to the Qt version of the OSK dialog when single line input is used (doesn't make much sense on multi-line text edits anyway)
- Adds more auto-repeat buttons to the OSK
- Properly resets auto-repeat when a different button was pressed

Megamouse commented:
implemented OSK password mode

kd-11 commented on rpcs3/Emu/RSX/Overlays/overlay_controls.h:
```diff
@@ -1095,10 +1095,19 @@ namespace rsx
+			bool m_reset_caret_pulse = false;
+			bool password_mode = false;
+
+			std::u32string value;
```
Why introduce a duplicate variable when the base class has a text variable for this very purpose?

kd-11 replied:
Nvm, I missed the password mode inclusion.

Megamouse replied:
That. and also there would be no way to really discriminate if a text is a placeholder without some extra logic anyway.

Megamouse replied:
Another thing worth mentioning is that the old code allowed to move the cursor while the placeholder was shown.
This was obviously wrong and is now fixed, since the caret logic works with the value member instead of the text member.

kd-11 commented on rpcs3/Emu/RSX/Overlays/overlay_edit_text.cpp:
```diff
@@ -172,7 +205,7 @@ namespace rsx
 					// TODO: Scrolling by using scroll offset
 					cmd.config.clip_region = true;
-					cmd.config.clip_rect   = {f32(x), f32(y), f32(x + w), f32(y + h)};
+					cmd.config.clip_rect   = {static_cast<f32>(x), static_cast<f32>(y), static_cast<f32>(x + w), static_cast<f32>(y + h)};
```
There's a lot of this static cast style change going around, let's have it as a separate commit.

kd-11 commented on rpcs3/Emu/RSX/Overlays/overlay_osk.cpp:
```diff
@@ -343,24 +345,22 @@ namespace rsx
 			{
 				const auto current_index = (start_index + count);
 				ensure(current_index <= index_limit);
+				++count;
```
:+1:

kd-11 commented on rpcs3/Emu/RSX/Overlays/overlay_osk.cpp:
```diff
@@ -568,6 +568,15 @@ namespace rsx
 				on_osk_input_entered();
 			}
 
+			if (m_preview.value.empty())
```
I feel like a simple ternary here would save multiple lines.

Megamouse replied:
fixed

kd-11 commented on rpcs3/Emu/RSX/Overlays/overlay_osk.cpp:
```diff
@@ -641,12 +649,6 @@ namespace rsx
 		{
 			m_preview.erase();
-
-			if (m_preview.text.empty())
```
It was implemented intentionally to restore some kind of hint text if completely empty. Not text to be returned, but like how textboxes and search bars work in modern UIs. If I remove all the text from this github comment text area, hint text saying 'Leave a comment' will appear every time.

Megamouse replied:
No. this was actually my addition that caused the regression of the attached issue.
The original code only showed the placeholder initially.
If you look at the set_text overrides in "overlay_edit_text" then you will see that It now properly handles the exact behavior you are describing.
Since I moved the logic to the text edit, we can now also use it in other future applications natively.

kd-11 commented on rpcs3/Emu/RSX/Overlays/overlay_controls.h:
```diff
@@ -468,15 +468,15 @@ namespace rsx
 			}
 
-			void set_text(const std::u32string& text)
+			virtual void set_wide_text(const std::u32string& text)
```
calling unicode strings wide is a bit old fashioned, maybe we can call it something else?
set_unicode_text() or set_text_unicode() or something.
std::wstring exists that literally means wide string, but that is not what we're using here, we're explicitly using utf-32.

Megamouse replied:
fixed

kd-11 commented on rpcs3/Emu/System.cpp:
```diff
@@ -300,10 +300,8 @@ void Emulator::Init(bool add_only)
 					sys_log.fatal("Failed to fix save data: %s (%s)", pending, fs::g_tls_error);
 					continue;
 				}
-				else
```
Yep, these formatting changes definitely need to go into their own commit at the very least.

Megamouse replied:
done

Megamouse commented:
Added some tiny QoL improvement.
- Added more auto-repeat buttons to the OSK (cursor movement, enter, space, backspace)
- Properly reset auto-repeat when a different button was pressed

Megamouse commented on rpcs3/Emu/RSX/Overlays/overlays.cpp:
```diff
@@ -152,16 +155,19 @@ namespace rsx
 										// the d-pad button was not pressed before, so this is a new button press
 										timestamp[pad_index] = steady_clock::now();
 										initial_timestamp[pad_index] = timestamp[pad_index];
+										last_pressed_button[pad_index] = button_id;
```
ah. I should probably unify this logic with the non repeat buttons

Megamouse commented:
hmm. I need to rethink this real quick.
don't mind the last commit for now

Megamouse commented:
Ok. The auto-repeat now feels good to me.
There's some random factor for next auto-repeat after releasing a button when you are already pressing 2 or more auto-repeat buttons.
Ideally I'd give the right for auto-repeat to the button that was initially pressed after all others, but I don't think that it's worth creating a button press hierarchy for such a misuse.

Megamouse commented:
fixed some typo in the last commit
